### ISSUE TITLE
Module export can't be named "module"

### ISSUE THREAD

**Issue by JCB-K** [vercel/next.js#26691]
### What version of Next.js are you using?

11.0.1

### What version of Node.js are you using?

12.20.1

### What browser are you using?

Chrome

### What operating system are you using?

OSX

### How are you deploying your application?

n/a

### Describe the Bug

If a module has a named export called `module`, or a wildcard import with the name `module` is used, Next.js crashes with the error:

> TypeError: Cannot read property 'data' of undefined

Example code:

```js
// file1.js
export const module = { value: 123 };

// file2.js
import { module } from './file1';
```

OR:

```js
// file1.js
export const anything = { value: 123 };

// file2.js
import * as module from './file1';
```

This behaviour doesn't happen in Webpack, which leads me to believe its a Next.js specific issue. However I'm wondering if it could be a Node-specific issue, but I haven't experienced it elsewhere.

### Expected Behavior

It should be possible to call an export `module`. 

### To Reproduce

1. clone this repo: https://github.com/JCB-K/import-as-module
2. yarn
3. yarn dev

The specific line that breaks is https://github.com/JCB-K/import-as-module/blob/master/pages/index.js

**Comment by mattcarlotta** [comment]
TLDR: Treat [module](https://nodejs.org/api/modules.html#modules_the_module_object) as a reserved word.

On a side note: I'm able to replicate a similar issue with the [create-react-app](https://create-react-app.dev/docs/getting-started). 

Some of it has to deal with [react-refresh](https://www.npmjs.com/package/react-refresh), some of it has to deal with `module` already being defined within the global scope and/or within an argument of a functions scope and some of it has to deal with Next expecting `module` not to be overridden. 

When declaring `module` within a file, you're essentially overriding the global/function scope `module`. This results in internal errors since: A.) Next appears to interpolate node's `module` to assign it as `data` to the `window` and/or B.) You can't declare a variable that is already defined in a global scope/functional scope with `const`.

Node's `module`:
```js
module Module {
  id: '.',
  path: '/home/m6d/Desktop/example',
  exports: {},
  parent: null,
  filename: '/home/m6d/Desktop/example/app.js',
  loaded: false,
  children: [],
  paths: [
    '/home/m6d/Desktop/example/node_modules',
    '/home/m6d/Desktop/node_modules',
    '/home/m6d/node_modules',
    '/home/node_modules',
    '/node_modules'
  ]
}
```
Next's `module`:
```js
{ children: [], exports: [Setter] };
```
Your `module`:
```js
import { module } from "../module";

console.log("🟡", module); // 🟡: module { value: 123 };
```

---

### Simple explanation regarding scope

Run this ES6 code (error):
```js
(function(module){
  console.log("module", module);
  const module = 123;  // Error: Identifier 'module' has already been declared`
  console.log("module", module);
)(0)
```
or
```js
var module = 0;
console.log("module", module);

const module = 123; // Error: Identifier 'module' has already been declared`
console.log("module", module);
```

Run this ES5 code (no error):
```js
(function(module){
  console.log("module", module); // 0
  var module = 123;
  console.log("module", module); // 123
})(0);
```
or 
```js
var module = 0;
console.log("module", module); // 0

var module = 123;
console.log("module", module); // 123
```



### In-depth explanation why it may or may not work

Depending on your `browserslist`, Webpack may compile the source code to ES6+ (error):

```js
(function(module, __webpack_exports__, __webpack_require__) {

    "use strict";
    __webpack_require__.r(__webpack_exports__);

    __webpack_require__.d(__webpack_exports__, "module", function() {
        return module;
    });

    const module = { // <<< "module" is already declared above; therefore you can't use "const"
        value: 123
    };
}),
```

or if `browserslist` supports older browser versions, then the code gets compiled for ES5 (no error):
```js
(function(module, __webpack_exports__, __webpack_require__) {

    "use strict";
    __webpack_require__.r(__webpack_exports__);

    __webpack_require__.d(__webpack_exports__, "module", function() {
        return module;
    });

    var module = { // <<< var overrides the "module" argument; therefore no error is thrown
        value: 123
    };
}),
```

The code above can work in production (for both CRA/Next) because it gets mangled in production:
```js
(self.webpackChunk_N_E = self.webpackChunk_N_E || []).push([
    [405], {
        4800: function(n, u, e) {
            "use strict";
            e.r(u), e.d(u, {
                default: function() {
                    return t
                }
            });
            var f = e(5893),
                r = 123; // <<< mangled "module"

            function t() {
                return (0, f.jsx)(f.Fragment, {
                    children: r
                })
            }
        },
        8581: function(n, u, e) {
            (window.__NEXT_P = window.__NEXT_P || []).push(["/", function() {
                return e(4800)
            }])
        }
    },
    function(n) {
        n.O(0, [774, 888, 179], (function() {
            return u = 8581, n(n.s = u);
            var u
        }));
        var u = n.O();
        _N_E = u
    }
]);
```

**Comment by JCB-K** [comment]
Cool, thanks for your thorough explanation! I figured it would be something along those lines, but couldn't figure out the details.

Would it be possible to throw a clearer error message? As it is right now it's quite hard to figure out what's wrong, and I had to do quite a bit of debugging to find the culprit.

**Comment by JCB-K** [comment]
Fair enough!

> On a related note, if you were to rename const module = { ... } to const module1 = { ... }, then you wouldn't run into any namespace issues.

Yup, that was my workaround here 👍

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
